lib/handlers/events/delivery/goals/k8s/KubernetesGoalLauncher.ts

```diff
@@ -0,0 +1,329 @@
+ */
+
+import {
+    automationClientInstance,
+    AutomationContextAware,
+    configurationValue,
+    doWithRetry,
+    logger,
+} from "@atomist/automation-client";
+import {
+    ExecuteGoalResult,
+    GoalInvocation,
+    GoalLauncher,
+} from "@atomist/sdm";
+import * as k8s from "@kubernetes/client-node";
+import * as cluster from "cluster";
+import * as _ from "lodash";
+import * as os from "os";
+import { loadKubeConfig } from "./k8config";
+
+/**
+ * GoalLauncher implementation that runs SDM goals inside k8s jobs.
+ *
+ * It reuses the podSpec of the deployed SDM to create a new jobSpec from.
+ * Subclasses may change the spec and job creation behavior by overwriting beforeCreation
+ * and/or afterCreation methods.
+ */
+export class KubernetesGoalLauncher implements GoalLauncher {
+
+    constructor(private readonly options: { isolateAll?: boolean } = { isolateAll: false }) {
+        this.init();
+    }
+
+    public async supports(gi: GoalInvocation): Promise<boolean> {
+        return !process.env.ATOMIST_ISOLATED_GOAL &&
+            (
+                // Goal is marked as isolated and SDM is configured to use k8s jobs
+                (gi.goal.definition.isolated && process.env.ATOMIST_GOAL_LAUNCHER === "kubernetes") ||
+                // Force all goals to run isolated via env var
+                process.env.ATOMIST_GOAL_LAUNCHER === "kubernetes-all" ||
+                // Force all goals to run isolated via explicit option
+                (this.options.isolateAll && process.env.ATOMIST_GOAL_LAUNCHER === "kubernetes") ||
+                // Force all goals to run isolated via explicit configuration
+                _.get(gi.configuration, "sdm.kubernetes.isolateAll", false) === true
+            );
+    }
+
+    public async launch(gi: GoalInvocation): Promise<ExecuteGoalResult> {
+        const { goalEvent, context } = gi;
+
+        // Using new ATOMIST_POD_NAME to overwrite the host name default
+        // This is to prevent breakage when users still have old ATOMIST_DEPLOYMENT_NAME env var defined
+        const podName = process.env.ATOMIST_POD_NAME || os.hostname();
+        const podNs = process.env.ATOMIST_DEPLOYMENT_NAMESPACE || "default";
+        const goalName = goalEvent.uniqueName.split("#")[0].toLowerCase();
+
+        const kc = loadKubeConfig();
+        const apps = kc.makeApiClient(k8s.Core_v1Api);
+        const batch = kc.makeApiClient(k8s.Batch_v1Api);
+
+        let podSpec: k8s.V1Pod;
+        try {
+            podSpec = (await apps.readNamespacedPod(podName, podNs)).body;
+        } catch (e) {
+            logger.error(`Failed to obtain parent pod spec from k8s: ${e.message}`);
+            return { code: 1, message: `Failed to obtain parent pod spec from k8s: ${e.message}` };
+        }
+
+        const jobSpec = createJobSpecWithAffinity(podSpec, goalEvent.goalSetId);
+
+        jobSpec.metadata.name = `${podSpec.spec.containers[0].name}-job-${goalEvent.goalSetId.slice(0, 7)}-${goalName}`;
```

Comment:
I would have expected that you would use `podName.split("-").slice(0, -2).join("-")` instead of `podSpec.spec.containers[0].name`.